File: MontyHall.py
import sys
import numpy as np

LENGTH = 100000

def monty_hall(length=LENGTH):
    arganswer = np.random.randint(0, 3, length)

    select = np.random.randint(0, 3, length)
    show = np.random.randint(0, 3, length)
    mask = (show == select) | (show == arganswer)
    while np.any(mask):
        show[mask] = np.random.randint(0, 3, np.count_nonzero(mask))
        mask = (show == select) | (show == arganswer)

    not_changing = np.count_nonzero(select == arganswer)
    changing = length - not_changing

    print(f'not changing:\t{np.round(not_changing/length, decimals=3)}')
    print(f'changing:\t{np.round(changing/length, decimals=3)}')

if __name__ == '__main__':
    print('Monty Hall Problem')
    if len(sys.argv) > 1:
        length = int(sys.argv[1])
    else:
        length = LENGTH

    monty_hall(length)

File: MontyHall.cpp
#include <iostream>
#include <cstdlib>
#include <ctime>

int randint(int min, int max);
void MontyHall(int& not_changing, int& changing);

int main(int argc, char* argv[]) {
    srand(time(0));  // Seed initialization

    int length = 10000;

    if (argc > 1) {
        length = std::stoi(argv[1]);
    }
    int not_changing = 0, changing = 0;

    for (int i = 0; i < length; i++) {
        MontyHall(not_changing, changing);
    }

    float ncResult = static_cast<float>(not_changing) / length;
    float cResult = static_cast<float>(changing) / length;

    std::cout << "Monty Hall Problem" << std::endl;
    std::cout << "not changing:\t" << ncResult << std::endl;
    std::cout << "changing:\t" << cResult << std::endl;

    return 0;
}

int randint(int min, int max) {
    return rand() % (max - min + 1) + min;
}

void MontyHall(int& not_changing, int& changing) {
    int answer = randint(0, 2);
    int select = randint(0, 2);
    int show = randint(0, 2);

    while (show == answer || show == select) {
        show = randint(0, 2);
    }

    if (select == answer) {
        not_changing++;
    } else {
        changing++;
    }
}
